staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/established_controller.go

```diff
@@ -0,0 +1,116 @@
+limitations under the License.
+*/
+
+package apiserver
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/internalclientset"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/workqueue"
+)
+
+// EstablishingController controls how and when CRD is established.
+type EstablishingController struct {
+	client *internalclientset.Clientset
+	queue  workqueue.DelayingInterface
+}
+
+// NewEstablishingController creates new EstablishingController.
+func NewEstablishingController(client *internalclientset.Clientset) *EstablishingController {
+	return &EstablishingController{
+		client: client,
+		queue:  workqueue.NewDelayingQueue(),
+	}
+}
+
+func (ec *EstablishingController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer ec.queue.ShutDown()
+
+	glog.Infof("Starting EstablishingController")
+	defer glog.Infof("Shutting down EstablishingController")
+
+	// only start one worker thread since its a slow moving API
+	go wait.Until(ec.runWorker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (ec *EstablishingController) runWorker() {
+	for ec.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.
+// It returns false when it's time to quit.
+func (ec *EstablishingController) processNextWorkItem() bool {
+	key, quit := ec.queue.Get()
+	if quit {
+		return false
+	}
+	defer ec.queue.Done(key)
+
+	err := ec.establishCRD(key.(string))
+	if err == nil {
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with: %v", key, err))
+	ec.queue.Add(key)
+
+	return true
+}
+
+// EstablishCRD is used to turn CRDs into the Established state.
+func (ec *EstablishingController) establishCRD(key string) error {
+	// Poll because if multiple servers try to change CRD (e.g. in HA) this can fail.
+	// That failure is normal, and so, we're trying to set it to Established again.
+	err := wait.Poll(500*time.Millisecond, 3*time.Second, func() (bool, error) {
+		crd, _ := ec.client.Apiextensions().CustomResourceDefinitions().Get(key, v1.GetOptions{})
+
+		if !apiextensions.IsCRDConditionTrue(crd, apiextensions.Established) &&
+			apiextensions.IsCRDConditionTrue(crd, apiextensions.Installing) {
+			establishedCondition := apiextensions.CustomResourceDefinitionCondition{
+				Type:    apiextensions.Established,
+				Status:  apiextensions.ConditionTrue,
+				Reason:  "InitialNamesAccepted",
+				Message: "the initial names have been accepted",
+			}
+
+			// Update storage with new CRD condition.
+			apiextensions.SetCRDCondition(crd, establishedCondition)
+
+			// Update server with new CRD condition.
+			var err error
+			_, err = ec.client.Apiextensions().CustomResourceDefinitions().UpdateStatus(crd)
+			if err != nil {
```

Comment:
this need a check for `errors.IsConflict`.

Actually, instead of polling, why not add the element back into the queue. Then we get throttling for free and don't block the queue (note: it's only one go routine processing the queue).